### PR TITLE
fix(git-url): clean up state of parsed URL

### PR DESCRIPTION
**Pull request by lcobucci**
## Changes

This ensures the state of `parsedUrl` doesn't have authentication credentials prior to appending the authentication tokens.

## Context

This is meant to verify and solve the issue described [here](https://github.com/renovatebot/renovate/discussions/24077#discussioncomment-6942191).

## Documentation

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/url.spec.ts [hunk 1]
```diff
@@ -77,6 +77,14 @@ describe('util/git/url', () => {
         'https://x-access-token:token@github.com/some/repo'
       );
     });
+
+    it('removes username/password from URL', () => {
+      expect(getHttpUrl('https://user:password@foo.bar/')).toBe('https://foo.bar/');
```
Comment (Shegox) [inline lib/util/git/url.spec.ts:82]: The url parse seems to take the `user` into the parsed url if no repo is specified
```suggestion
      expect(getHttpUrl('https://user:password@foo.bar/someOrg/someRepo')).toBe('https://foo.bar/someOrg/someRepo');
```

File: lib/util/git/url.spec.ts [hunk 2]
```diff
@@ -77,6 +77,14 @@ describe('util/git/url', () => {
...
+    });
+    
+    it('replaces username/password with given token', () => {
+      expect(getHttpUrl('https://user:password@foo.bar/', 'another-user:a-secret-pwd')).toBe('https://another-user:a-secret-pwd@foo.bar/');
```
Comment (Shegox) [inline lib/util/git/url.spec.ts:86]: same as above
```suggestion
      expect(getHttpUrl('https://user:password@foo.bar/someOrg/someRepo', 'another-user:a-secret-pwd')).toBe('https://another-user:a-secret-pwd@foo.bar/someOrg/someRepo');
```

File: lib/util/git/url.ts [hunk 3]
```diff
@@ -15,6 +15,7 @@ export function getHttpUrl(url: string, token?: string): string {
     ? parsedUrl.protocol
     : 'https';
 
+  parsedUrl.user = parsedUrl.password = null;
   parsedUrl.token = token ?? '';
```
Comment (Shegox) [inline lib/util/git/url.ts:19]: While `parsedUrl` extracts the user + password it doesn't allow setting them. Just setting `user` to `''` should be enough.

```suggestion
  parsedUrl.user = '';
```

Comment (lcobucci) [reply lib/util/git/url.ts:19]: Based on your comment, the diff should be applied on L18, right?

```diff
-parsedUrl.user = parsedUrl.password = null;
+parsedUrl.user = '';
 parsedUrl.token = token ?? '';
```

Comment (Shegox) [reply lib/util/git/url.ts:19]: sorry, yeah. Added a suggestion at the right place.

File: lib/util/git/url.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ export function getHttpUrl(url: string, token?: string): string {
     ? parsedUrl.protocol
     : 'https';
 
+  parsedUrl.user = parsedUrl.password = null;
```
Comment (Shegox) [inline lib/util/git/url.ts:18]: While `parsedUrl` extracts the user + password it doesn't allow setting them. Just setting `user` to `''` should be enough.

```suggestion
  parsedUrl.user = '';
```

**Comment by lcobucci** [comment]
@Shegox it should be good now :)

**Comment by lcobucci** [comment]
@Shegox do you know if the broken tests are related to real expectations?

**Comment by lcobucci** [comment]
@Shegox do you know if the broken tests are related to real expectations?

**Comment by Shegox** [comment]
I'm not sure, I haven't seen the syntax with the username being part of the hostRule in the wild. 
As the test was added by @lfcyja in #23354, maybe they can add some insights. Thanks!

**Review by viceice — COMMENTED** [review]
needs prettier fix

**Comment by lcobucci** [comment]
> I'm not sure, I haven't seen the syntax with the username being part of the hostRule in the wild. As the test was added by @lfcyja in #23354, maybe they can add some insights. Thanks!

Looking at the patch, it doesn't seem strictly related - more like side effects of the status quo. @Shegox should we move forward with this? If so, I'll fix the test expectations and address the coding style issues.

**Review by Shegox — APPROVED** [review]
Looks good and for GitHub and GitLab this should work as expected. For the other platforms I personally don't see a reason why the "user" should be part of the URL in the `user@domain` syntax. 

@rarkins @viceice could you please take a look at this PR and how we should proceed. Thanks!

**Review by viceice — APPROVED** [review]

File: lib/util/git/url.spec.ts [hunk 5]
```diff
@@ -77,6 +77,14 @@ describe('util/git/url', () => {
         'https://x-access-token:token@github.com/some/repo'
```
Comment (8164933729) [inline lib/util/git/url.spec.ts:78]: Wyd

**Pull request merged**